opentelemetry-exporter-geneva/geneva-uploader/src/ingestion_service/uploader.rs

```diff
@@ -0,0 +1,212 @@
+    pub auth_info: IngestionGatewayInfo,
+    pub moniker: String,
+    pub config: GenevaUploaderConfig,
+    pub http_client: Client,
+    pub monitoring_endpoint: String,
+}
+
+impl GenevaUploader {
+    /// Constructs a GenevaUploader by calling the GenevaConfigClient
+    ///
+    /// # Arguments
+    /// * `config_client` - Initialized GenevaConfigClient
+    /// * `uploader_config` - Static config (namespace, event, version, etc.)
+    ///
+    /// # Returns
+    /// * `Result<GenevaUploader>` with authenticated client and resolved moniker/endpoint
+    #[allow(dead_code)]
+    pub(crate) async fn from_config_client(
+        config_client: &GenevaConfigClient,
+        uploader_config: GenevaUploaderConfig,
+    ) -> Result<Self> {
+        let (auth_info, moniker_info, monitoring_endpoint) =
+            config_client.get_ingestion_info().await?;
+
+        let http_client = Client::builder().timeout(Duration::from_secs(30)).build()?;
+
+        Ok(Self {
+            auth_info,
+            moniker: moniker_info.name,
+            config: uploader_config,
+            http_client,
+            monitoring_endpoint,
+        })
+    }
+
+    /// Creates the GIG upload URI with required parameters
+    #[allow(dead_code)]
+    fn create_upload_uri(&self, data_size: usize, event_name: &str, event_version: &str) -> String {
+        let now: DateTime<Utc> = Utc::now(); //TODO - this need to be calculated from the bond data
+        let end_time = now + ChronoDuration::minutes(5); //TODO - this need to be calculated from the bond data
+
+        // Format times in ISO 8601 format with fixed precision
+        // Using .NET compatible format (matches DateTime.ToString("O"))
+
+        let start_time = format!(
+            "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:07}Z",
+            now.year(),
+            now.month(),
+            now.day(),
+            now.hour(),
+            now.minute(),
+            now.second(),
+            now.nanosecond() / 100 // Convert nanoseconds to 7-digit precision
+        );
+
+        let end_time = format!(
+            "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:07}Z",
+            end_time.year(),
+            end_time.month(),
+            end_time.day(),
+            end_time.hour(),
+            end_time.minute(),
+            end_time.second(),
+            end_time.nanosecond() / 100 // Convert nanoseconds to 7-digit precision
+        );
+
+        // URL encode parameters
+        // TODO - Maintain this as url-encoded in config service to avoid conversion here
+        let encoded_monitoring_endpoint: String =
+            byte_serialize(self.monitoring_endpoint.as_bytes()).collect();
+
+        let encoded_source_identity: String =
+            byte_serialize(self.config.source_identity.as_bytes()).collect();
+
+        // Create a source unique ID - using a UUID to ensure uniqueness
+        let source_unique_id = Uuid::new_v4().to_string();
+
+        // Create the query string
+        format!(
```

Comment:
Since this would be called for every upload, maybe use `String::with_capacity()` and `push_str()` instead of using the `format!` macro for more efficient construction of the URL.

Reply:
I noticed that you used `write!` macro. Is that better than calling `push_str` to the pre-allocated string? If yes, then let's make that change for `GenevaConfigClient::new()` and `GenevaConfigClient::get_ingestion_info()` as well.